signature/internal/base/envelope.go

```diff
@@ -0,0 +1,213 @@
+	// validation before the core verify process.
+	if len(e.Raw) == 0 {
+		return nil, &signature.MalformedSignatureError{}
+	}
+
+	// core verify process.
+	content, err := e.Envelope.Verify()
+	if err != nil {
+		return nil, err
+	}
+
+	// validation after the core verify process.
+	if err = validateEnvelopeContent(content); err != nil {
+		return nil, err
+	}
+
+	return content, nil
+}
+
+// Content returns the validated signature information and payload.
+func (e *Envelope) Content() (*signature.EnvelopeContent, error) {
+	if len(e.Raw) == 0 {
+		return nil, &signature.MalformedSignatureError{Msg: "raw signature is empty"}
+	}
+
+	content, err := e.Envelope.Content()
+	if err != nil {
+		return nil, err
+	}
+
+	if err = validateEnvelopeContent(content); err != nil {
+		return nil, err
+	}
+
+	return content, nil
+}
+
+// validateSignRequest performs basic set of validations on SignRequest struct.
+func validateSignRequest(req *signature.SignRequest) error {
+	if err := validatePayload(&req.Payload); err != nil {
+		return err
+	}
+
+	if err := validateSigningTime(req.SigningTime, req.Expiry); err != nil {
+		return err
+	}
+
+	if req.Signer == nil {
+		return &signature.MalformedSignatureError{Msg: "signer is nil"}
+	}
+
+	_, err := req.Signer.KeySpec()
+	return err
+}
+
+// validateEnvelopeContent validates the content which includes signerInfo and
+// payload.
+func validateEnvelopeContent(content *signature.EnvelopeContent) error {
+	if err := validatePayload(&content.Payload); err != nil {
+		return err
+	}
+	return validateSignerInfo(&content.SignerInfo)
+}
+
+// validateSignerInfo performs basic set of validations on SignerInfo struct.
+func validateSignerInfo(info *signature.SignerInfo) error {
+	if len(info.Signature) == 0 {
+		return &signature.MalformedSignatureError{Msg: "signature not present or is empty"}
+	}
+
+	if info.SignatureAlgorithm == 0 {
+		return &signature.MalformedSignatureError{Msg: "SignatureAlgorithm is not present"}
+	}
+
+	signingTime := info.SignedAttributes.SigningTime
+	if err := validateSigningTime(signingTime, info.SignedAttributes.Expiry); err != nil {
+		return err
+	}
+
+	return validateCertificateChain(
+		info.CertificateChain,
+		signingTime,
+		info.SignatureAlgorithm,
+	)
+}
+
+// validateSigningTime checks that signing time is within the valid range of
+// time duration.
+func validateSigningTime(signingTime, expireTime time.Time) error {
+	if signingTime.IsZero() {
+		return &signature.MalformedSignatureError{Msg: "signing-time not present"}
+	}
+
+	if !expireTime.IsZero() && (expireTime.Before(signingTime) || expireTime.Equal(signingTime)) {
+		return &signature.MalformedSignatureError{Msg: "expiry cannot be equal or before the signing time"}
+	}
+	return nil
+}
+
+// validatePayload performs validation of the payload.
+func validatePayload(payload *signature.Payload) error {
+	switch payload.ContentType {
+	case signature.MediaTypePayloadV1:
+		if len(payload.Content) == 0 {
```

Comment:
Should we move this check outside the the switch as payload cannot be of length 0?

Reply:
As mentioned in community call, payload valiation will be moved to `notation-go`.

Reply:
to summarize: remove any MediaType related checks in core go.

Reply:
i see, so we keep the content size check in `notation-core-go`?

Reply:
Yes, lets keep content length check because if content is empty than what are we going to sign.